Add e2e test

danielvegamyhre opened the pull request:
Part of #46 

Currently only 1 test case:
- Create JobSet with DNS hostnames enabled and ensure pods can reach each other via pod hostnames

Reviewer Notes:
- For some reason, after the Kind cluster has been created, it takes a long time for the CRD to be installed, JobSet controller manager pods to become ready (~30 seconds), etc. so I had to add a sleep after the `jobset_deploy` function in `./hack/e2e-test.sh`, otherwise the resources required for the e2e test were not ready yet.
- The bash script I wrote to wait until it can successfully ping all pods by hostnames seems to take ~30 seconds before pings are successful, so I set the e2e test timeout as 60 seconds. However, I find this strange since all pods are supposedly in a running state, so not sure what is causing the delay before pings work.

ahg-g commented on hack/e2e-test.sh:
```diff
@@ -0,0 +1,28 @@
+}
+function jobset_deploy {
+    (cd config/manager && $KUSTOMIZE edit set image controller=$IMAGE_TAG)
+    kubectl apply -k test/e2e/config
```
shouldn't we do `--server-side`?

ahg-g commented on hack/e2e-test.sh:
```diff
@@ -0,0 +1,28 @@
+    $KIND load docker-image $IMAGE_TAG --name $KIND_CLUSTER_NAME
+}
+function jobset_deploy {
+    (cd config/manager && $KUSTOMIZE edit set image controller=$IMAGE_TAG)
```
can you make the shell script print every command it executes so we know what is going on?

ahg-g commented:
/hold

just in case someone lgtms by mistake

tenzen-y commented on Makefile:
```diff
@@ -44,6 +44,14 @@ INTEGRATION_TARGET ?= ./test/integration/...
 PROJECT_DIR := $(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
 
+E2E_TARGET ?= ./test/e2e/...
+E2E_KIND_VERSION ?= kindest/node:v1.23.12
```
Do we support v1.23? IIUC, since we use the CEL validation, our minimal supporting version is v1.25, right?

danielvegamyhre replied:
Makes sense, updated to v1.25.1

tenzen-y commented on Makefile:
```diff
@@ -184,7 +192,21 @@ GINKGO = $(shell pwd)/bin/ginkgo
+KIND = $(shell pwd)/bin/kind
+.PHONY: kind
+kind:
+	@GOBIN=$(PROJECT_DIR)/bin GO111MODULE=on $(GO_CMD) install sigs.k8s.io/kind@v0.16.0
```
Leyt's uses a newer kind version v0.18.0 :)

https://github.com/kubernetes-sigs/kind/releases/tag/v0.18.0

danielvegamyhre replied:
done

tenzen-y commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,177 @@
+// /*
+// Copyright 2023 The Kubernetes Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+// 	http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// */
```
```suggestion
/*
Copyright 2023 The Kubernetes Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

nit

Comments are doubly applied.

tenzen-y reviewed (commented):
Overall looks good :)
some nits.

ahg-g commented on test/e2e/suite_test.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+const (
+	timeout  = 1000 * time.Second
```
since this is a large interval, can you change to minutes?

danielvegamyhre replied:
Sorry this was a remnant from debugging, changed it back to 60 sec / 1 minute.

ahg-g commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,171 @@
+	) // end of DescribeTable
+}) // end of Describe
+
+func checkJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType) (bool, error) {
```
This and the two functions below should be moved to a common library shared with the integration tests; may be under test/util/util.go (I don't like the name, but I can't think of a better one :()

ahg-g commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,171 @@
+		expectCondition jobset.JobSetConditionType
+	}
+
+	ginkgo.DescribeTable("",
```
lets not use tables here, we should have few independent e2e tests.

ahg-g reviewed (commented):
we need to setup presubmit on this as well, right?

danielvegamyhre commented:
@ahg-g Thanks for the review, I addressed your comments and this is ready for another look. Also yes we will need presubmits for this, and I will set those up with a PR to test-infra next.

ahg-g commented:
/label tide/merge-method-squash

ahg-g commented:
/lgtm

danielvegamyhre commented:
/hold cancel
